theano/tensor/nnet/corr_gemm.c

```diff
@@ -172,50 +171,61 @@ PyArrayObject* corrMM(PyArrayObject* bottom,
     }
 
     // Create temporary columns
-    npy_intp col_dim[2];
-    col_dim[0] = (npy_intp)(nChannels * kW * kH);
-    col_dim[1] = (npy_intp)(topHeight * topWidth);
-    PyArrayObject* col = (PyArrayObject*)PyArray_EMPTY(2,
-		                           col_dim,
-                                           PyArray_TYPE(top),
-					   0);
+    const int max_threads = %(omp_max_threads)s < batchSize ? %(omp_max_threads)s : batchSize;
+
+    npy_intp col_dim[3];
+    col_dim[0] = (npy_intp)max_threads;
+    col_dim[1] = (npy_intp)(nChannels * kW * kH);
+    col_dim[2] = (npy_intp)(topHeight * topWidth);
+    PyArrayObject* col = (PyArrayObject*)PyArray_ZEROS(3,
+		                                          col_dim,
+                                        PyArray_TYPE(top),
+					                                   0);
     if (NULL == col)
     {
         PyErr_Format(PyExc_RuntimeError,
-                "CorrMM failed to allocate working memory of %%ld x %%ld\n",
-                col_dim[0], col_dim[1]);
+                "CorrMM failed to allocate working memory of"
+                " %%ld x %%ld x %%ld\n",
+                col_dim[0], col_dim[1], col_dim[2]);
         return NULL;
     }
 
     // Define some useful variables
     const int bottom_stride = PyArray_STRIDES(bottom)[0]/%(n_bytes)f;
     const int top_stride = PyArray_STRIDES(top)[0]/%(n_bytes)f;
-    const int K_ = col_dim[0];
-    const int N_ = col_dim[1];
+    const int K_ = col_dim[1];
+    const int N_ = col_dim[2];
+    const int col_stride = (K_ * N_);
     const int M_ = nFilters;
     const %(c_float_type)s one = 1.0;
     const %(c_float_type)s zero = 0.0;
     char NTrans = 'N';
     char Trans = 'T';
-
     PyArrayObject *output;
+
+    %(omp_set_threads)s(max_threads);
+    
     if (direction == 0) {  // forward pass
         output = top;
         // valid correlation: im2col, then gemm
         // Iterate over batch
-        for (int n = 0; n < batchSize; n++) {
+        %(omp_flags)s
```

Comment:
@nouiz, I did some timing in #3689 which I'll copy here. Basically, I found that it was best to use 1 core for BLAS and parallelize the batch across cores. I was working with fairly large batches (128) on a 12 core machine.

```
Timing done on 12-core Intel Xeon X5650 with 24GB RAM.
BLAS was MKL from Anaconda Python.

Image shape, filter shape
(128, 3, 128, 128), (96, 3, 5, 5)
Imagenet-like
MKL     OpenMP  Speedup  Notes
1       12      1        Legacy, OpenMP, baseline
1       1       .097     Legacy, single thread
1       1       .75      CorrMM, MKL only
2       1       1.3      “
4       1       1.9      “
8       1       2.0      “
12      1       1.7      “
1       2       1.5      CorrMM, MKL single thread + OpenMP
1       4       2.7      “
1       8       4.5      “
1       12      5.3      “

Image shape, filter shape
(128, 85, 1, 258), (64, 85, 1, 20)
Spectrogram-like
MKL     OpenMP  Speedup  Notes
1       12      1        Legacy, OpenMP, baseline
1       1       .095     Legacy, single thread
1       1       .50      CorrMM, MKL only
2       1       .78      “
4       1       1.2      “
8       1       1.5      “
12      1       1.3      “
1       2       .94      CorrMM, MKL single thread + OpenMP
1       4       2.0      “
1       8       4.0      “
1       12      5.2      “
```